File: letter-combinations-of-a-phone-number/letter-combinations-of-a-phone-number.java
class Solution {
    public void letter(List<String>l,String d,String p,int i,String []map)
    {
        if(i>=d.length())
        {
            l.add(p);
            return ;
        }
        String lee=map[d.charAt(i)-'0'];
        for(int i1=0;i1<lee.length();i1++)
        {
            letter(l,d,p+lee.charAt(i1),i+1,map);
        }
    }
    public List<String> letterCombinations(String digits) {
        List<String>l=new ArrayList<>();
        if(digits.length()==0)
            return l;
        String []map={"0","1","abc","def","ghi","jkl","mno","pqrs","tuv","wxyz"};
        letter(l,digits,"",0,map);
        return l;
    }
}

File: letter-combinations-of-a-phone-number/letter-combinations-of-a-phone-number.scala
object Solution {
    def num2letter(d:String) = {
        d match {
            case  "2"=> "abc"
            case  "3"=> "def"
            case  "4"=> "ghi"
            case  "5"=> "jkl"
            case  "6"=> "mno"
            case  "7"=> "pqrs"
            case  "8"=> "tuv"
            case  "9"=> "wxyz"
            case _ => ""
        }
    }
    def letterCombinations(digits: String): List[String] = {
        val letters = digits.split("").map(num2letter).toList
        letters.foldLeft(List(""))( (l, x) => for (li <- l; xi<-x) yield {li + xi} )  
    }
}